README.md
# Encoder / Decoder in javascript for QOI the quite okay image format.
Inspired by the work of [Francis Stokes](https://github.com/LowLevelJavaScript/QOI) and his encoder implementation
on [Youtube](https://www.youtube.com/watch?v=GgsRQuGSrc0).

# Reference
QOI format [https://qoiformat.org/](https://qoiformat.org/) by [Dominic Szablewski](http://twitter.com/phoboslab).

# Example
```node
'use strict';

const
    {encode, decode, EncodeStream, DecodeStream} = require('qoi-image'),
    {createHash} = require('crypto'),
    {createReadStream, readFileSync} = require('fs'),
    {join} = require('path'),
    filename = 'testcard_rgba',
    infile = join(__dirname, '..', 'reference', `${filename}.bin`),
    outfile = join(__dirname, '..', 'reference', `${filename}.qoi`),
    binBuffer = readFileSync(infile),
    qoiBuffer = readFileSync(outfile),
    refBinHash = createHash('sha1').update(binBuffer).digest('hex'),
    refQoiHash = createHash('sha1').update(qoiBuffer).digest('hex'),
    encoded = encode(binBuffer, 256, 256, 4, 1),
    decoded = decode(qoiBuffer),
    encodeHash = createHash('sha1').update(encoded).digest('hex'),
    decodeHash = createHash('sha1').update(decoded).digest('hex');

console.log(`Reference BIN hash: ${refBinHash}`);
console.log(`Reference QOI hash: ${refQoiHash}`);

console.log(`Encode hash: ${encodeHash}`);
console.log(`Decode hash: ${decodeHash}`);

createReadStream(infile)
    .pipe(new EncodeStream(256, 256, 4, 1))
    .pipe(createHash('sha1').setEncoding('hex'))
    .on('finish', function() {
        const hash = this.read();
        console.log(`EncodeStream hash: ${hash}`);
    });

createReadStream(outfile)
    .pipe(new DecodeStream())
    .pipe(createHash('sha1').setEncoding('hex'))
    .on('finish', function() {
        const hash = this.read();
        console.log(`DecodeStream hash: ${hash}`);
    });
```

CHANGELOG.md
### 1.1.1
- fix readme

### 1.1.0
- added EncodeStream and DecodeStream
- added changelog
- updated Readme
- fixed typo

### 1.0.0
- initial version

tests/index.js
/* eslint-disable no-console */
'use strict';

const
    {createHash} = require('crypto'),
    {createReadStream, readFileSync} = require('fs'),
    {join} = require('path'),
    {encode, decode, EncodeStream, DecodeStream} = require('./../src'),
    filename = 'testcard_rgba',
    infile = join(__dirname, '..', 'reference', `${filename}.bin`),
    outfile = join(__dirname, '..', 'reference', `${filename}.qoi`),
    binBuffer = readFileSync(infile),
    qoiBuffer = readFileSync(outfile),
    refBinHash = createHash('sha1').update(binBuffer).digest('hex'),
    refQoiHash = createHash('sha1').update(qoiBuffer).digest('hex'),
    encoded = encode(binBuffer, 256, 256, 4, 1),
    decoded = decode(qoiBuffer),
    encodeHash = createHash('sha1').update(encoded).digest('hex'),
    decodeHash = createHash('sha1').update(decoded).digest('hex');

console.log(`Reference BIN hash: ${refBinHash}`);
console.log(`Reference QOI hash: ${refQoiHash}`);

console.log(`Encode hash: ${encodeHash}`, refQoiHash === encodeHash ? 'match' : 'no match');
console.log(`Decode hash: ${decodeHash}`, refBinHash === decodeHash ? 'match' : 'no match');

createReadStream(infile)
    .pipe(new EncodeStream(256, 256, 4, 1))
    .pipe(createHash('sha1').setEncoding('hex'))
    .on('finish', function() {
        const hash = this.read();
        console.log(`EncodeStream hash: ${hash}`, refQoiHash === hash ? 'match' : 'no match');
    });

createReadStream(outfile)
    .pipe(new DecodeStream())
    .pipe(createHash('sha1').setEncoding('hex'))
    .on('finish', function() {
        const hash = this.read();
        console.log(`DecodeStream hash: ${hash}`, refBinHash === hash ? 'match' : 'no match');
    });


src/Decode.js
'use strict';

const QOI = require('./QoiConstants');

module.exports = (buffer) => {
    const
        data = new DataView(ArrayBuffer.isView(buffer) ? buffer.buffer : buffer),
        dataSize = data.byteLength,
        seenPixels = Array.from({length: 64}, () => 0);

    if (dataSize < QOI.HEADER_SIZE + QOI.TAIL_SIZE) {
        return null;
    }

    const
        header = data.getUint32(0, false),
        width = data.getUint32(4, false),
        height = data.getUint32(8, false),
        channels = data.getUint8(12),
        cs = data.getUint8(13);

    if (width < 1 || height < 1 || channels < 3 || channels > 4 || cs < 0 || cs > 1 || header !== QOI.HEADER_MAGIC) {
        return null;
    }

    const
        len = width * height * channels,
        bytes = new DataView(new ArrayBuffer(len)),
        lastPixel = dataSize - QOI.TAIL_SIZE;

    for (let color = 0x000000FF, offset = 0, run = 0, index = 14; offset < len; offset += channels) {
        if (run > 0) {
            run--;
        } else if (index < lastPixel) {
            const
                b1 = data.getUint8(index++),
                b1Masked = b1 & QOI.MASK_2;

            if (b1 === QOI.OP_RGB) {
                color = data.getUint16(index, false) << 16 | data.getUint8(index + 2) << 8 | color & 0xFF;
                index += 3;
            } else if (b1 === QOI.OP_RGBA) {
                color = data.getUint32(index, false);
                index += 4;
            } else if (b1Masked === QOI.OP_INDEX) {
                color = seenPixels[b1];
            } else if (b1Masked === QOI.OP_DIFF) {
                color = (((color >> 24) + (((b1 >> 4) & 0x03) - 2)) & 0xFF) << 24
                    | (((color >> 16) + (((b1 >> 2) & 0x03) - 2)) & 0xFF) << 16
                    | (((color >> 8) + ((b1  & 0x03) - 2)) & 0xFF) << 8
                    | color & 0xFF;
            } else if (b1Masked === QOI.OP_LUMA) {
                const
                    b2 = data.getUint8(index++),
                    dg = (b1 & 0x3F) - 32;

                color = ((color >> 24) + (dg - 8 + ((b2 >> 4) & 0x0F)) & 0xFF) << 24
                    | (((color >> 16) + dg) & 0xFF) << 16
                    | (((color >> 8) + (dg - 8 + (b2 & 0x0F))) & 0xFF) << 8
                    | color & 0xFF;
            } else if (b1Masked === QOI.OP_RUN) {
                run = b1 & 0x3F;
            }

            const hash = ((color >> 24 & 0xFF) * 3
                + (color >> 16 & 0xFF) * 5
                + (color >> 8 & 0xFF) * 7
                + (color & 0xFF) * 11) % 64;

            seenPixels[hash] = color;
        }

        if (channels === 4) {
            bytes.setUint32(offset, color, false);
        } else {
            bytes.setUint8(offset, color >> 24 & 0xFF);
            bytes.setUint8(offset + 1, color >> 16 & 0xFF);
            bytes.setUint8(offset + 2, color >> 8 & 0xFF);
        }
    }

    return new Uint8Array(bytes.buffer);
}

src/Encode.js
'use strict';

const QOI = require('./QoiConstants');

module.exports = (buffer, width, height, channels, colorspace) => {
    const
        len = width * height * channels,
        lastPixel = buffer.byteLength - channels,
        maxSize = width * height * (channels + 1) + QOI.HEADER_SIZE + QOI.TAIL_SIZE,
        data = new DataView(ArrayBuffer.isView(buffer) ? buffer.buffer : buffer),
        bytes = new DataView(new ArrayBuffer(maxSize)),
        seenPixels = Array.from({length: 64}, () => 0);

    bytes.setUint32(0, QOI.HEADER_MAGIC, false);
    bytes.setUint32(4, width, false);
    bytes.setUint32(8, height, false);
    bytes.setUint8(12, channels);
    bytes.setUint8(13, colorspace);

    let index = 14;

    for (let color, prevColor = 0x000000FF, run = 0, offset = 0; offset < len; offset += channels) {
        if (channels === 3) {
            color = ((data.getUint16(offset, false) << 16) | (data.getUint8(offset + 2)) << 8) | 0xFF;
        } else {
            color = data.getUint32(offset, false);
        }

        (color === prevColor) && run++;

        if ((color === prevColor && (run === 62 || offset === lastPixel)) || (color !== prevColor && run)) {
            bytes.setUint8(index++, QOI.OP_RUN | run - 1);
            run = 0;
        }

        if (color !== prevColor) {
            const hash = ((color >> 24 & 0xFF) * 3
                + (color >> 16 & 0xFF) * 5
                + (color >> 8 & 0xFF) * 7
                + (color & 0xFF) * 11) % 64;

            if (color === seenPixels[hash]) {
                bytes.setUint8(index++, QOI.OP_INDEX | hash);
            } else {
                seenPixels[hash] = color;

                if ((color & 0xFF) === (prevColor & 0xFF)) {
                    const
                        dr = ((color >> 24 & 0xFF) - (prevColor >> 24 & 0xFF)) << 24 >> 24,
                        dg = ((color >> 16 & 0xFF) - (prevColor >> 16 & 0xFF)) << 24 >> 24,
                        db = ((color >> 8 & 0xFF) - (prevColor >> 8 & 0xFF)) << 24 >> 24,
                        drg = dr - dg,
                        dbg = db - dg;

                    if (dr > -3 && dr < 2 && dg > -3 && dg < 2 && db > -3 && db < 2) {
                        bytes.setUint8(index++, QOI.OP_DIFF | dr + 2  << 4 | dg + 2 << 2 | db + 2);
                    } else if (dg > -33 && dg < 32 && drg > -9 && drg < 8 && dbg > -9 && dbg < 8) {
                        bytes.setUint16(index, (QOI.OP_LUMA | dg + 32) << 8 | drg + 8 << 4 | dbg + 8, false);
                        index += 2;
                    } else {
                        bytes.setUint32(index, QOI.OP_RGB << 24 | color >> 8 & 0xFFFFFF, false);
                        index += 4;
                    }
                } else {
                    bytes.setUint32(index, QOI.OP_RGBA << 24 | color >> 8, false);
                    bytes.setUint8(index + 4, color & 0xFF);
                    index += 5;
                }
            }
        }
        prevColor = color;
    }

    bytes.setBigUint64(index, QOI.TAIL_MAGIC, false);

    return new Uint8Array(bytes.buffer.slice(0, index + QOI.TAIL_SIZE));
}


reference/decode.c
#define QOI_IMPLEMENTATION
#include "qio.h"
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>

void print_usage() {
    printf("Usage: decode -i file -o file -b num [3 - 4]\n");
    exit(EXIT_FAILURE);
}

int main (int argc, char **argv) {
    int c;
    int option_index = 0;
    int bpp = -1;
    char *infile = "";
    char *outfile = "";

    while (1) {
        static struct option long_options[] = {
            {"in",  required_argument, 0, 'i'},
            {"out", required_argument, 0, 'o'},
            {"bpp", required_argument, 0, 'b'},
            {0, 0, 0, 0}
        };

        c = getopt_long (argc, argv, ":i:o:b:", long_options, &option_index);

        if (c == -1) {
            break;
        }

        switch (c) {
            case 'i':
                infile = optarg;
                break;
            case 'o':
                outfile = optarg;
                break;
            case 'b':
                bpp = atoi(optarg);
                break;
            case '?':
            case ':':
            default:
                print_usage();
                abort ();
                return 1;
        }
    }

    if (infile == "" || outfile == "" || bpp < 3 || bpp > 4) {
        print_usage();
        return 1;
    }

    if (access(infile, F_OK | R_OK) != 0) {
        puts("Failure reading from input file");
        return 1;
    }

    qoi_desc imgDesc;

    void* decodedImage = qoi_read(infile, &imgDesc, bpp);
    int decodedLength = imgDesc.width * imgDesc.height * bpp;

    FILE* fp = fopen(outfile, "wb");
    fwrite(decodedImage, 1, decodedLength, fp);
    fclose(fp);

    printf("Wrote decoded QOI image (%i bytes)\n", decodedLength);

    return 0;
}

reference/encode.c
#define QOI_IMPLEMENTATION
#include "qio.h"
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>

void print_usage() {
    printf("Usage: encode -i file -o file -w num -h num -b num [3 - 4] -c num [0 - 1]\n");
    exit(EXIT_FAILURE);
}

int main (int argc, char **argv) {
    int c;
    int option_index = 0;
    int width = -1;
    int height = -1;
    int bpp = -1;
    int cp = 1;
    char *infile = "";
    char *outfile = "";

    while (1) {
        static struct option long_options[] = {
            {"in",         required_argument, 0, 'i'},
            {"out",        required_argument, 0, 'o'},
            {"width",      required_argument, 0, 'w'},
            {"height",     required_argument, 0, 'h'},
            {"bpp",        required_argument, 0, 'b'},
            {"colorspace", required_argument, 0, 'c'},
            {0, 0, 0, 0}
        };

        c = getopt_long (argc, argv, ":i:o:w:h:b:c:", long_options, &option_index);

        if (c == -1) {
            break;
        }

        switch (c) {
            case 'i':
                infile = optarg;
                break;
            case 'o':
                outfile = optarg;
                break;
            case 'w':
                width = atoi(optarg);
                break;
            case 'h':
                height = atoi(optarg);
                break;
            case 'b':
                bpp = atoi(optarg);
                break;
            case 'c':
                cp = atoi(optarg);
                break;
            case '?':
            case ':':
            default:
                print_usage();
                abort ();
                return 1;
        }
    }

    if (infile == "" || outfile == "" || width < 1 || height < 1 || bpp < 3 || bpp > 4 || cp < 0 || cp > 1) {
        print_usage();
        return 1;
    }

    if (access(infile, F_OK | R_OK) != 0) {
        puts("Failure reading from input file");
        return 1;
    }

    qoi_desc imgDesc = {
        .width      = width,
        .height     = height,
        .channels   = bpp,
        .colorspace = cp
    };

    FILE *fp = fopen(infile, "rb");
    void* buffer = malloc(imgDesc.width * imgDesc.height * imgDesc.channels);
    fread(buffer, imgDesc.width * imgDesc.height * imgDesc.channels, 1, fp);
    fclose(fp);

    int encodedLength = qoi_write(outfile, buffer, &imgDesc);

    printf("Wrote QOI image (%i bytes)\n", encodedLength);
    free(buffer);

    return 0;
}

src/EncodeStream.js
'use strict';

const
    {Transform} = require('stream'),
    QOI = require('./QoiConstants');

class EncodeStream extends Transform
{
    #firstChunk;
    #length;
    #cache;
    #bpp;
    #header;
    #run;
    #prev;
    #buffer;
    #lastPixel;
    #pos;

    constructor (width, height, channels, colorspace, opts)
    {
        const {highWaterMark} = typeof opts === 'object' ? opts || {} : {};
        super({highWaterMark});

        this.#firstChunk = true;
        this.#length = width * height * channels;
        this.#lastPixel = this.#length - channels;
        this.#cache = Array.from({length: 64}, () => 0);
        this.#bpp = channels;
        this.#header = Buffer.allocUnsafe(14);
        this.#prev = 0x000000FF;
        this.#buffer = Buffer.allocUnsafe(0);
        this.#run = 0;
        this.#pos = 0;

        this.#header.writeUInt32BE(QOI.HEADER_MAGIC, 0);
        this.#header.writeUInt32BE(width, 4);
        this.#header.writeUInt32BE(height, 8);
        this.#header.writeUInt8(channels, 12);
        this.#header.writeUInt8(colorspace, 13);
    }

    _encode (chunk)
    {
        let
            size = 0,
            data = Buffer.allocUnsafe(1 + chunk.byteLength + chunk.byteLength / this.#bpp);

        for (let col, offset = 0; offset < chunk.byteLength; offset += this.#bpp, this.#pos += this.#bpp) {
            if (this.#bpp === 3) {
                col = chunk.readUInt16BE(offset) << 16 | chunk.getUint8(offset + 2) << 8 | 0xFF;
            } else {
                col = chunk.readUInt32BE(offset);
            }

            const match = col === this.#prev;

            match && this.#run++;

            if ((match && (this.#run === 62 || this.#pos === this.#lastPixel)) || (!match && this.#run)) {
                data.writeUInt8(QOI.OP_RUN | this.#run - 1, size++);
                this.#run = 0;
            }

            if (!match) {
                const hash = ((col >> 24 & 0xFF) * 3
                    + (col >> 16 & 0xFF) * 5
                    + (col >> 8 & 0xFF) * 7
                    + (col & 0xFF) * 11) % 64;

                if (col === this.#cache[hash]) {
                    data.writeUInt8(QOI.OP_INDEX | hash, size++);
                } else {
                    this.#cache[hash] = col;

                    if ((col & 0xFF) === (this.#prev & 0xFF)) {
                        const
                            dr = ((col >> 24 & 0xFF) - (this.#prev >> 24 & 0xFF)) << 24 >> 24,
                            dg = ((col >> 16 & 0xFF) - (this.#prev >> 16 & 0xFF)) << 24 >> 24,
                            db = ((col >> 8 & 0xFF) - (this.#prev >> 8 & 0xFF)) << 24 >> 24,
                            drg = dr - dg,
                            dbg = db - dg;

                        if (dr > -3 && dr < 2 && dg > -3 && dg < 2 && db > -3 && db < 2) {
                            data.writeUInt8(QOI.OP_DIFF | dr + 2  << 4 | dg + 2 << 2 | db + 2, size++);
                        } else if (dg > -33 && dg < 32 && drg > -9 && drg < 8 && dbg > -9 && dbg < 8) {
                            data.writeUInt8(QOI.OP_LUMA | dg + 32, size++);
                            data.writeUInt8(drg + 8 << 4 | dbg + 8, size++);
                        } else {
                            data.writeUInt8(QOI.OP_RGB, size);
                            data.writeUInt8(col >> 24 & 0xFF, size + 1);
                            data.writeUInt8(col >> 16 & 0xFF, size + 2);
                            data.writeUInt8(col >> 8 & 0xFF, size + 3);
                            size += 4;
                        }
                    } else {
                        data.writeUInt8(QOI.OP_RGBA, size);
                        data.writeUInt8(col >> 24 & 0xFF, size + 1);
                        data.writeUInt8(col >> 16 & 0xFF, size + 2);
                        data.writeUInt8(col >> 8 & 0xFF, size + 3);
                        data.writeUInt8(col & 0xFF, size + 4);
                        size += 5;
                    }
                }
            }
            this.#prev = col;
        }
        return data.slice(0, size);
    }

    _transform(chunk, encoding, callback)
    {
        if (this.#firstChunk) {
            this.#firstChunk = false;
            this.push(this.#header);
        }

        this.#buffer = Buffer.concat([this.#buffer, chunk]);
        const len = this.#buffer.byteLength - this.#buffer.byteLength % this.#bpp;

        if (this.#buffer.byteLength % this.#bpp === 0) {
            this.push(this._encode(this.#buffer.slice(0, len)));
        }

        this.#buffer = this.#buffer.slice(len);

        callback(null);
    }

    _flush(callback)
    {
        const tail = Buffer.allocUnsafe(QOI.TAIL_SIZE);

        tail.writeBigInt64BE(QOI.TAIL_MAGIC);
        callback(null, tail);
    }
}

module.exports = EncodeStream;